### ISSUE TITLE
flagstat machine readable output format

### ISSUE THREAD

**Issue by SichongP** [samtools/samtools#1077]
#### Is your feature request related to a problem? Please specify.
I'm trying to consolidate flagstat output for multiple bam files in my workflow and generate a report with a clean table for all stats. I could achieve this with a simple script and regex but obviously this is prone to failing should there be any updates in the output formatting for flagstat in the future. As such I'm wondering if there is a way to make flagstat output a more standardized and machine readable format (a csv or yaml perhaps)?
If not, may I request such a feature? 
#### Describe the solution you would like.
An option for flagstat when specified output will be in a csv or yaml format that can be easily consolidated.

**Comment by tseemann** [comment]
This would be useful.  I would choose TSV over CSV, and JSON over YAML, but I am all for this.

**Comment by daviesrob** [comment]
Well, it's output hasn't changed for years (four to be precise) but it wouldn't be hard to make it print other formats.

I too prefer JSON to YAML.

**Comment by raivivek** [comment]
If no one's working on this feature yet, I'd like to take a shot at it. I'd also find it personally useful. Let me know if that's alright.

**Comment by daviesrob** [comment]
@raivivek Feel free to have a go, it shouldn't be that difficult.

I think you'd want to use `--output-fmt` for the long option, with maybe `-O` (capital O) for the short option.  So you would need to change `SAM_OPT_GLOBAL_OPTIONS` to:
```c
SAM_OPT_GLOBAL_OPTIONS('-', 0, 'O', '-', '-', '@')
```
and also add `case 'O':` to the options switch statement to handle the format selection.  It's important to do that because `parse_sam_global_opt()` doesn't understand formats like `json`, so you need to bypass it for this option.

It would be best to move the `printf` statements out of `bam_flagstat()` so you can make a separate function for each output format.  You could also turn them into `fprintf` so you can have a `-o` or `--output` option to set the output file name.

Note that samtools code uses four spaces for indentation.  Tabs and trailing spaces are not allowed.  Also, don't forget to add tests and update the `samtools-flagstat.1` manual page.

**Comment by raivivek** [comment]
Great, thanks for the details. That gives me a good start.

**Comment by raivivek** [comment]
@daviesrob Also, I was wondering. Do you have any specific debugging practices that you use? I'm using the usual "change -> make -> run/test", but can't seem to find any general development instructions. Thanks!

**Comment by raivivek** [comment]
Okay, I think have a draft working. We can discuss how do we want to display existing stats in the desired format. For TSV, I have below:
```
91957 	0 	total (QC-passed reads + QC-failed reads)
0 	0 	secondary
0 	0 	supplementary
0 	0 	duplicates
91957 	0 	mapped (100.00% : N/A)
91957 	0 	paired in sequencing
45996 	0 	read1
45961 	0 	read2
91957 	0 	properly paired (100.00% : N/A)
91957 	0 	with itself and mate mapped
0 	0 	singletons (0.00% : N/A)
0 	0 	with mate mapped to a different chr
0 	0 	with mate mapped to a different chr (mapQ>=5)
```

Not sure of the ideal way to do it for JSON format.

**Comment by raivivek** [comment]
@SichongP Thanks for the idea. I've referenced the pull request where I have ongoing changes related to this issue. Feel free to comment on the code/here itself and I'll push changes accordingly. The current changes are just some defaults I used to test if it works.

**Comment by raivivek** [comment]
@SichongP Pull request was merged recently. You might wish to checkout the latest version and feel free to close/comment.

**Comment by valeriuo** [comment]
Closed by #1106
